src/fal/cli/dbt_runner.py

```diff
@@ -166,3 +160,117 @@ def _create_fal_result_file(target_path: str, run_index: int):
         shutil.copy(
             fal_run_result, os.path.join(target_path, f"fal_results_{run_index}.json")
         )
+
+
+def _process_logs(raw_logs: str):
+    logs, output = [], []
+    for line in raw_logs.splitlines():
+        output.append(line)
+        try:
+            json_line = json.loads(line)
+        except json.JSONDecodeError:
+            logger.error(line.rstrip())
+        else:
+            if "message" in json_line:
+                try:
+                    actual_msg = json.loads(json_line["message"])["msg"]
+                except:
+                    actual_msg = json_line["message"]
+            else:
+                actual_msg = line.rstrip()
+            logger.info(actual_msg)
+
+    return logs, output
+
+
+@contextmanager
+def _patch_log_manager(log_manager, output):
+    # Temporarily patch the log_manager's stdout and stderr with the given
+    # 'output' buffer, and revert it back at the exit of the context manager.
+    original_stdout, original_stderr = log_manager.stdout, log_manager.stderr
+    try:
+        log_manager.set_output_stream(output)
+        yield
+    finally:
+        log_manager.reset_handlers()
+        log_manager.set_output_stream(original_stdout, original_stderr)
+
+
+def _dbt_run_through_python(args: List[str], target_path: str, run_index: int):
+    import io
+    from dbt.main import handle_and_check
+    from dbt.logger import log_manager
+
+    # This is the core function where we execute the dbt run command. We'll initially
+    # redirect all possible streams into our output (this is very similiar how subprocess)
+    # is doing.
+    output = io.StringIO()
+    with redirect_stdout(output), redirect_stderr(output), _patch_log_manager(
+        log_manager, output
+    ):
+        # Then we'll run the given DBT command, and get the 'run_results'.
+        try:
+            run_results, success = handle_and_check(args)
+        except BaseException as exc:
+            run_results = None
+            return_code = getattr(exc, "code", 1)
+            traceback.print_exc(exc)
+        else:
+            return_code = 0 if success else 1
+
+    logger.debug(f"dbt exited with return code {return_code}")
+
+    raw_output = output.getvalue()
+    logs, _ = _process_logs(raw_output)
+
+    # The 'run_results' object has a 'write()' method which is basically json.dump().
+    # We'll dump it directly to the fal results file (instead of first dumping it to
+    # run results and then copying it over).
+    if run_results is not None:
+        run_results.write(os.path.join(target_path, f"fal_results_{run_index}.json"))
+    else:
+        raise NotImplementedError("TODO: What should happen if dbt run fails and there are no run results file?")
+
+    return return_code, raw_output, logs
+
+
+def dbt_run_through_python(
+    args: argparse.Namespace, models_list: List[str], target_path: str, run_index: int
+):
```

Comment:
return type here, at least to all public methods (still deciding if we should add a linter rule about this)